### ISSUE TITLE
can we incorporate reaction-field hybrid factory into `perses`?

### ISSUE THREAD

**Issue by dominicrufa** [choderalab/perses#1130]
Can we incorporate [this reaction field hybrid system factory](https://github.com/dominicrufa/aludel/blob/7d5e348a9b5a6b7dd01b85bb3c35c064777f4d80/aludel/atm.py#L756) into the full perses stack? it is not too crappy in comparison to the current vanilla factory at the moment:
(vanilla factory)
![image](https://user-images.githubusercontent.com/46943710/201404213-312e21e4-4fa8-4793-bbf6-53c7c5cbeacb.png)
(rf)
![image](https://user-images.githubusercontent.com/46943710/201404301-48b41d2d-94ae-48da-bdf7-8140ac782838.png)

**Comment by zhang-ivy** [comment]
Can you explain why you wrote a new factory from scratch rather than using one of the existing factories `HybridTopologyFactory` or `RESTCapableHybridTopologyFactory` as the base class? 

At this point we have 3 completely different factories.. we probably want to avoid having a completely different factory for every hybrid system variant as this will be difficult to maintain

**Comment by dominicrufa** [comment]
> Can you explain why you wrote a new factory from scratch rather than using one of the existing factories HybridTopologyFactory or RESTCapableHybridTopologyFactory as the base class?

I kind of made an on-the-spot judgement call to get this done as quickly as possible with the fewest code footprint.

1. there's a lot of auxiliary information and function calls in the `__init__` method of the base hybrid factory that is unnecessary for a softcored rxn field hybrid factory, which would have necessitated overwriting the exisitng base class init anyway.
2. each valence force has its own specific method for translation to hybrid; it's easier to write a single function that translates bonds, angles, torsions to hybrid (esp exclusively as a custom force). it removes redundancy and it helped me to understand what exact valence interactions were happening immensely. 
3. the SH reaction field nonbonded force uses a different expression for exclusions that involves the vacuum partial charges as well as their scaled chargeprod, so that required different `CustomBondForce` expressions (as well as a different number of forces and different meanings of those forces). also, i needed to completely remove the `NonbondedForce` in favour of a `CustomNonbondedForce`.
4. i wanted to incorporate the endstate energy tests directly into the factory as a method, which is not trivial to do/subclass easily given the current validations tests.
5. I wrote a PME -> SH reaction field converter class for endstate systems anyway (for jax-md), and that already provided the energy expressions for pairwise nonbonded/exception nonbonded terms, so it was easier to just directly subclass that.

> we probably want to avoid having a completely different factory for every hybrid system variant as this will be difficult to maintain

I agree. unfortunately, the re-learning curve to try to figure out how the base factory works and what i needed to do to modify it was steep, and even after i did re-learn how it works, I realized that I would have easily modded >1k lines to make it work, which turned out to be ~2x more than the number of lines to write the new one altogether. 

I don't know if there is a good general fix to this. hybrid system creation logic and treatment is complicated in the single topology regime; i found that writing the most abstract functionality for force groups more generally minimized the amount of code needed.

**Comment by ijpulidos** [comment]
> At this point we have 3 completely different factories (though `RESTCapableHybridTopologyFactory` does subclass `HybridTopologyFactory` for some of its helper functions) .. we probably want to avoid having a completely different factory for every hybrid system variant as this will be difficult to maintain

I agree with this approach, as possible. FYI, we do have an issue opened for this kind of factory-like pattern in https://github.com/choderalab/perses/issues/1051 

I guess it boils down to figuring out if such a pattern is actually reasonable, since as @dominicrufa mentions, there may not be a big (if any) overlap between these.

**Comment by ijpulidos** [comment]
@dominicrufa Does it make sense to ask ourselves if instead of rebasing your class on the current ones, maybe we can rewrite the current ones using your approach and still have a factory pattern?

**Comment by dominicrufa** [comment]
totally; it's just a time commitment.

**Comment by jchodera** [comment]
Just to clarify the strategy I hope we can follow here: 
* **Develop a general API for relative free energy calculations for `openmmtools.alchemy`.** We ultimately want to migrate all alchemical factories to `openmmtools`, using a single common API to host all alchemical factories.
* **Refactor this class to separate `Force` objects that compute alchemical interactions from those that do not.** In order to take advantage of the `openmmtools.multistate` sampler acceleration for energy matrix computation, we need to separate out the small amount of work that computes lambda-dependent interactions into separate `Force` objects.
* **Streamline the code.** We make it as simple (but modular) as we can.
* **Document how it works.** We should make it as clear as possible to understand.

Once this is done, my hope is that the software scientists can refactor the existing approaches in `perses` into `openmmtools.alchemy` following a similar strategy, which may give us opportunity to streamline these factories as we do so and end up with a long-term maintenance codebase within `openmmtools`.
